Frame.cc

```diff
@@ -1061,3 +1075,73 @@ bool Frame::GetSpectralData(std::vector<float>& data, casacore::SubImage<float>&
     }
     return data_ok;
 }
+
+bool Frame::GetRegionSpectralData(std::vector<std::vector<double>>& stats_values, int region_id, int profile_index,
+    int profile_stokes) {
+    int delta_channels = 10; // the increment of channels for each step
+    int dt_target = 75; // the target time elapse for each step, in the unit of milliseconds
+    int profile_size = NumChannels(); // total number of channels
+    auto& region = _regions[region_id];
+    // get statistical requirements for this process
+    CARTA::SetSpectralRequirements_SpectralConfig config;
+    if (!region->GetSpectralConfig(config, profile_index)) {
+        return false;
+    }
+    int stats_size = config.stats_types().size();
+    std::vector<int> requested_stats(config.stats_types().begin(), config.stats_types().end());
+    // initialize the size of statistical results
+    std::vector<std::vector<double>> results(stats_size);
+    for (int i = 0; i < stats_size; ++i) {
+        results[i].resize(profile_size);
+    }
+    // get region state for this process
+    RegionState region_state = region->GetRegionState();
+    // get statistical profile data
+    int start = 0;
+    int count, end;
+    casacore::SubImage<float> sub_image;
+    while (start < profile_size) {
+        // start the timer
+        auto tStart = std::chrono::high_resolution_clock::now();
+        // check if frontend queries changed, if so, terminate this loop process
+        if (!_connected) {
```

Comment:
To make it easier to reuse this sequence of checks from the swizzled calculation function on the HDF5 loader, maybe refactor it into its own public function and pass the frame object as a parameter to the loader function so that you can call the function on the frame from inside the loader.

Inside the loader function I guess you would check this before reading each X slice from the swizzled dataset.

Reply:
You could call the same function to check state from here and from the loader function.